### PR TITLE
Adding support for webpack externals.

### PR DESCRIPTION
**Pull request by internettrans**
Resolves issue #18.

### CODE REVIEW

**Comment by internettrans** [comment]
You can view the output bundle externals.js at https://gist.github.com/joeldenning/6cf9492eabaf4967fa200c361696cf2b

File: esm-webpack-plugin.js
Comment (purtuga) [inline esm-webpack-plugin.js]: What is request by the time this code is hit?

`externals` has several interfaces - I actually found the project where I use it with a combination of Object and function.

File: esm-webpack-plugin.js
Comment (purtuga) [inline esm-webpack-plugin.js]: The [docs specifically indicate](https://webpack.js.org/configuration/externals/#externalstype) that this value should match that of `output.libraryTarget`. Is it really safe to change it here?

**Review by purtuga — COMMENTED** [review]
First - Thank you again for your time and contribution 🙏 

So now that I have had a chance to catch up on `externals` and the intent of of this PR, I really like this. It would allow one to have much smaller bundles and just bring in the `externals` as `import ` statements.  This is especially nice for Web Components (my initial purpose for this plugin) which could then bring in `imports` (other Components) from a CDN in order to be able to compose more involved UI components. 

I do have a few concerns - the main one being that this would likely be a backwards-breaking change. Currently, the changes made with this PR assume that if `externals` are defined, that automatically implies that they must be `import`'ed in the module, which ignores the many other use cases where users might not want that (ex. in my use case where I use `externals`, I used it where I load code async and then execute that code by providing it a `context` with all the runtime dependencies in it).  

I'm also thinking that we introduce a Plugin option that must be set to `true` in order to enable this behaviour as well as document the behavior along with limitations if any. 


## A few Ideas for discussion

What do you think of these:

### Only process if a specific external value is defined
Ideally (IMO) the `webpack.config` should drive/trigger this addition of imports in this plugin. Example: If the `externals` value was defined like an `import` statement, then this plugin would add it to the top of the output bundle:

```javascript
{
    externals: {
        foo: 'import  * as foo from "some/location"'
    }
}
```
And this plugin can then detect that the value for the external is `import ...` and just take that verbatim and add it to the output bundle.  What I don't know is if Webpack will like that syntax. Although `externals` has an `externalTypes` property, it seems that [some of those values can also be used inline when defining an external as a string](https://webpack.js.org/configuration/externals/#string), so my example above actually cause webpack to process it as a dynamic import (they already have an `externalTypes` value of `import`).  I guess for the purpose of this plugin's functionality, we could come up with our own syntax (`_import * as foo from 'some/location"' - ugly, I know 😄 )
 

### Define external "overrides" in the plugin's options
We could also define a "simpler" approach to ensuring the user has control over the `import` by adding an options to the plugin that allows them to define how the import is set. In this approach, the syntax could actually mirror the idea I suggested above:
```javascript
module.exports = {
    //...
    plugins: [
        new EsmWebpackPlugin({
            externalImports: {
                foo: 'import  * as foo from "some/location"',
                bar: 'import * as bar from "https://cdn.jsdelivr.net/npm/bar@1.0.0/bar.min.js"'
            }
        })
    ]
}
```
In this approach, I actually don't think we probably need the full `import ...` statement - maybe just the path/URL from there is should be loaded, the rest as you PR show can be obtained from the code implementation.

Thanks again and sorry for the delayed review/response. 

/Paul

**Comment by internettrans** [comment]
Hey there, thanks for the review.

> I do have a few concerns - the main one being that this would likely be a backwards-breaking change.

This is true. Before these changes, externals were assumed to be global variables. After this change, externals are assumed to be ES modules. [This gist](https://gist.github.com/joeldenning/0d2869f08f98ddc1fb16a1fb4da3c511) shows how externals were compiled before my changes, and [this gist](https://gist.github.com/joeldenning/6cf9492eabaf4967fa200c361696cf2b) shows how externals are compiled afterwards.

> Currently, the changes made with this PR assume that if externals are defined, that automatically implies that they must be import'ed in the module, which ignores the many other use cases where users might not want that

I didn't fully understand your explanation of how you've used externals in the past, so there's a chance I'm missing something here. My understanding of webpack externals is that they mark certain modules as runtime dependencies instead of build-time dependencies. According to [these webpack docs](https://webpack.js.org/configuration/externals/#root):

> The externals configuration option provides a way of excluding dependencies from the output bundles. Instead, the created bundle relies on that dependency to be present in the consumer's (any end-user application) environment.

In the case of an ESM environment, dependencies are expressed via `import` statements. I do not see how it could mean anything else. Here's what webpack does with other libraryTargets externals:

| libraryTarget | compiled externals |
| ------------ | ------------------- |
| `amd` | `define(['dep1', 'dep2'], (dep1, dep2) => {...})` |
| `cjs` | `const dep1 = require('dep1'); const dep2 = require('dep2');` |
| `system` | `System.register(['dep1', 'dep2'], ...)` |
| `esm` | `import * as dep1 from 'dep1'; import * as dep2 from 'dep2';` |

Another clue here is that rollup implements ESM externals via `import` statements.

> I'm also thinking that we introduce a Plugin option that must be set to true in order to enable this behaviour as well as document the behavior along with limitations if any.

Since the definition of externals is well understood and has existed for >5 years, I don't see any value in hiding it behind an option. I doubt many users are depending on the current behavior of global variables, so probably few users would be impacted. However, it is a breaking change so probably a major version bump of this plugin is probably the right thing to do. 

> We could also define a "simpler" approach to ensuring the user has control over the import by adding an options to the plugin that allows them to define how the import is set. In this approach, the syntax could actually mirror the idea I suggested above:

Webpack has a [well-established, official way](https://webpack.js.org/configuration/externals/#object) of defining externals, including renaming the module request string to a runtime import path. I don't see any reason to deviate from it.

File: esm-webpack-plugin.js
Comment (internettrans) [reply esm-webpack-plugin.js]: `m.request` is either a string or an object. See https://github.com/webpack/webpack/blob/cd4af164784f3938353102179adfcbcf8f53949d/lib/SystemMainTemplatePlugin.js#L46

File: esm-webpack-plugin.js
Comment (internettrans) [reply esm-webpack-plugin.js]: Yes. See the following lines of code, which show the webpack implementation:

- https://github.com/webpack/webpack/blob/cd4af164784f3938353102179adfcbcf8f53949d/lib/ExternalModule.js#L136
- https://github.com/webpack/webpack/blob/cd4af164784f3938353102179adfcbcf8f53949d/lib/ExternalModule.js#L89
- https://github.com/webpack/webpack/blob/cd4af164784f3938353102179adfcbcf8f53949d/lib/AmdMainTemplatePlugin.js#L49

I implemented this for the SystemMainTemplatePlugin in https://github.com/webpack/webpack/pull/8864. SystemJS also works with ES module objects, and he suggested that SystemJS use the same implementation for externals as AMD. Changing the externalType changes how webpack compiles it. The externalType for `var` or `assign` assumes that externals are global variables, which is undesireable for ESM.

**Comment by purtuga** [comment]
Sorry for the confusing example I mentioned in my prior comment. I was only meaning to describe the case where the ESM module may actually want to use a Global (ex. something in `window.*` or `global.*`, as detailed here:

>The bundle with external dependencies can be used in various module contexts, such as CommonJS, AMD, global and ES2015 modules. The external library may be available in any of these forms:
>
> - **root:** The library should be available as a global variable (e.g. via a script tag).

(ref: https://webpack.js.org/configuration/externals/#externals )

In this Global case where `root` is used, the changes you are suggesting would prevent a user of this plugin from achieving that. Here is an example of a (simplified) webpack config I have in one of my private projects:

```javascript
config.externals = {
  foo: '../../test/fixtures/externals/foo-external.js',
  bar: '../../test/fixtures/externals/bar-external.js',
  ext: 'root MY_GLOBAL_VAR',
}
```

The goal here is to output a bundle that looks something like this (see module 2 below - I used your PR and changed a few things in the test):

```javascript
import * as __WEBPACK_EXTERNAL_MODULE__0__ from '../../test/fixtures/externals/foo-external.js';
import * as __WEBPACK_EXTERNAL_MODULE__1__ from '../../test/fixtures/externals/bar-external.js';


var LIB =
/******/ (function(modules) { // webpackBootstrap
// removed outptup for brevity 
/******/ })
/************************************************************************/
/******/ ([
/* 0 */
/***/ (function(module, exports) {

module.exports = __WEBPACK_EXTERNAL_MODULE__0__;

/***/ }),
/* 1 */
/***/ (function(module, exports) {

module.exports = __WEBPACK_EXTERNAL_MODULE__1__;

/***/ }),
/* 2 */
/***/ (function(module, exports) {

module.exports = MY_GLOBAL_VAR;

/***/ }),
/* 3 */
/***/ (function(module, __webpack_exports__, __webpack_require__) {

"use strict";
__webpack_require__.r(__webpack_exports__);
/* harmony export (binding) */ __webpack_require__.d(__webpack_exports__, "externals", function() { return externals; });
/* harmony import */ var foo__WEBPACK_IMPORTED_MODULE_0__ = __webpack_require__(0);
/* harmony import */ var foo__WEBPACK_IMPORTED_MODULE_0___default = /*#__PURE__*/__webpack_require__.n(foo__WEBPACK_IMPORTED_MODULE_0__);
/* harmony reexport (module object) */ __webpack_require__.d(__webpack_exports__, "foo", function() { return foo__WEBPACK_IMPORTED_MODULE_0__; });
/* harmony import */ var bar__WEBPACK_IMPORTED_MODULE_1__ = __webpack_require__(1);
/* harmony import */ var bar__WEBPACK_IMPORTED_MODULE_1___default = /*#__PURE__*/__webpack_require__.n(bar__WEBPACK_IMPORTED_MODULE_1__);
/* harmony import */ var ext__WEBPACK_IMPORTED_MODULE_2__ = __webpack_require__(2);
/* harmony import */ var ext__WEBPACK_IMPORTED_MODULE_2___default = /*#__PURE__*/__webpack_require__.n(ext__WEBPACK_IMPORTED_MODULE_2__);

const externals = {bar: bar__WEBPACK_IMPORTED_MODULE_1__, EXT: ext__WEBPACK_IMPORTED_MODULE_2__["EXT"]};

/***/ })
/******/ ]);

LIB === undefined && console.error('esm-webpack-plugin: nothing exported!');
const _LIB$foo = LIB['foo'];
const _LIB$externals = LIB['externals'];

export {
    _LIB$foo as foo,
    _LIB$externals as externals
}
```

This why I was suggesting an approach to detecting when you actually want to use a Global -vs- converting the external to an `import` (I wonder how Webpack will solve that in the next version 🤔 )


I still would like to get this change in - even if there is no clear path to distinguishing between when one might want a true global -vs- conversion to `import`, but would like to do so behind a plugin option so that users of this plugin can control the behavior to fit their needs.  If we make the default of that flag to disable this functionality by default, then I can see releasing this in a minor version - otherwise it would be a major. 

/Paul

**Comment by internettrans** [comment]
ESM code that references globals should reference them via `const thing = window.externalVarName`, not via `import thing from 'externalVarName'`, since no bundler or browser supports the latter. RollupJS' current ESM implementation and webpack's upcoming support for ESM both operate this way. esm-webpack-plugin is a bit weird in that it instructs users to set the libraryTarget to `var` instead of `module`. The official implementation of ESM instructs users to set the libraryTarget to `module` and then externals will be treated as ES imports. A more complete implementation of this esm-webpack-plugin would have been to add the `module` libraryTarget instead of using `var` libraryTarget.

One thing to note is that esm-webpack-plugin already deviates from libraryTarget `var` in a substantial way because it exposes the library via `export` instead of `window.globalVar =` (which is how `var` exports are implemented). So the project has already embraced a weird/broken usage of libraryTarget `var`, and my code just fully embraces that the libraryTarget is essentially `module`, not `var`.

**Comment by purtuga** [comment]
Thanks for the continued dialog in this Joel. Yes, this plugin is a bit
weird (more than a bit 🙂) and a hack until official support lands. It was
never meant to mimic the official upcoming implementation by Webpack and my
plan is to quickly deprecate it after webpack releases support of ESM.

I appreciate your efforts in attempting to mirror the ESM official
implementation with the introduction for externals support. Sometime this
weekend (time permitting) I will merge this PR in and will proceed to put
it behind a plugin option that must be set in order for this behavior to be
enabled. You are welcome to use the plugin and enable that option for your
use cases. You are also welcome to fork this repo and created a more
standards compliant plugin - no hard feelings from me :) - and honestly it
sounds like you are a lot more plugged into how bundlers work than me.

Thanks again for your time on this.


-- 
Paul T. -- Sent from Mobile

**Comment by internettrans** [comment]
Fair enough - I don't see the purpose of the option or its default, but would rather work together on this package than maintain a fork. What brings me to esm-webpack-plugin is that I am the maintainer of single-spa, the most popular microfrontends framework, and am exploring options for server-side rendering. One option I'm looking into is using esm on the server side, in combination with [nodejs loaders](https://nodejs.org/dist/latest-v13.x/docs/api/esm.html#esm_experimental_loaders). Specifically, I am experimenting with [this import map loader](https://github.com/node-loader/node-loader-import-maps) and [this http loader](https://github.com/node-loader/node-loader-http) that allow for loading modules from across the network via bare specifiers. That approach only works with ESM, and many single-spa users are using webpack, which brought me to esm-webpack-plugin. If I can get it all working and feel confident about it, I may add esm-webpack-plugin to single-spa's CLI and documentation for those using webpack 4.

There's one remaining thing with esm externals and webpack that I am trying to figure out - Webpack assumes that all external modules are CJS by default, and only treats them as ES modules if they have the `__esModule` flag set. Real NodeJS ESM modules do not have that flag set, so webpack is currently treating all the ESM externals as CJS, which messes up how it interprets default / named exports on the module. We had the same issue when implementing libraryTarget `system` and were able to work around it, but that workaround won't work for real NodeJS ES modules since those are non-configurable.

^ All this just to say that hopefully I find a workaround for it, and if I do I'll send in another PR for it.

**Comment by internettrans** [comment]
I had a few minutes just now so I added that plugin option. I called it `moduleExternals` and have it defaulting to false. I also added a test verifying that global externals still work when the option is not turned on.

File: test/utils.js
Comment (purtuga) [inline test/utils.js:55]: This is awsome - thank you

**Review by purtuga — APPROVED** [review]
Thanks again for your contribution.

**Comment by purtuga** [comment]
published `@purtuga/esm-webpack-plugin@1.4.0` to npm - https://www.npmjs.com/package/@purtuga/esm-webpack-plugin

**Pull request merged**
